url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/util/http/TestForHttpRequestUri.java

```diff
@@ -0,0 +1,191 @@
+	}
+
+	public void testDecodeScheme() throws IOException {
+		{
+			MarkableFileReader reader = new MarkableFileReader("http".getBytes());
+			assertEquals("http", HttpGetRequestUri.scheme(reader));
+			reader.close();
+		}
+
+		{
+			MarkableFileReader reader = new MarkableFileReader("https".getBytes());
+			assertEquals("https", HttpGetRequestUri.scheme(reader));
+			reader.close();
+		}
+
+		{
+			MarkableFileReader reader = new MarkableFileReader("htt".getBytes());
+			try {
+			assertEquals("htt", HttpGetRequestUri.scheme(reader));
+			assertTrue(false);
+			} catch(IOException e) {
+				
+			}
+			assertEquals(0, reader.getFilePointer());
+		}
+	}
+
+	public void testEncode001() throws IOException {
+		HttpGetRequestUri uri = new HttpGetRequestUri("*");
+		CashKyoroFile output = new CashKyoroFile(512);
+		uri.encode(output.getLastOutput());
+		output.seek(0);
+		byte[] buffer = new byte[(int)output.length()];
+		int len = output.read(buffer);
+		String tag = new String(buffer, 0, len);
+		assertEquals("*", tag);
+	}
+
+	public void testEncode002() throws IOException {
+		HttpGetRequestUri uri = new HttpGetRequestUri("/announce");
+		uri.putVale("01", "value01");
+
+		CashKyoroFile output = new CashKyoroFile(512);
+		uri.encode(output.getLastOutput());
+		output.seek(0);
+		byte[] buffer = new byte[(int)output.length()];
+		int len = output.read(buffer);
+		String tag = new String(buffer, 0, len);
+		assertEquals("/announce?01=value01", tag);
+	}
+
+	public void testEncode003() throws IOException {
+		HttpGetRequestUri uri = new HttpGetRequestUri("/announce");
+		uri.putVale("01", "value01");
+		uri.putVale("02", "value02");
+		uri.putVale("03", "value03");
+		CashKyoroFile output = new CashKyoroFile(512);
+		uri.encode(output.getLastOutput());
+		output.seek(0);
+		byte[] buffer = new byte[(int)output.length()];
+		int len = output.read(buffer);
+		String tag = new String(buffer, 0, len);
+		assertEquals("/announce?01=value01&02=value02&03=value03", tag);
+	}
+
+	public void testDecode001() throws IOException {
+		CashKyoroFile base = new CashKyoroFile(512);
+		base.addChunk("*".getBytes());
+		MarkableFileReader reader = new MarkableFileReader(base, 512);
+
+		HttpGetRequestUri value = HttpGetRequestUri.decode(reader);
+		assertEquals("*", value.getPath());
+	}
+
+	public void testDecode002() throws IOException {
+		CashKyoroFile base = new CashKyoroFile(512);
+		base.addChunk("/announce?01=value01".getBytes());
+		MarkableFileReader reader = new MarkableFileReader(base, 512);
+
+		HttpGetRequestUri value = HttpGetRequestUri.decode(reader);
+		assertEquals("/announce", value.getPath());
+		Iterator<String> keys = value.keySet().iterator();
+		assertEquals("/announce", value.getPath());
+		String key = keys.next();
+		assertEquals("01", key);
+		assertEquals("value01", value.getValue(key));
+		assertEquals(1, value.keySet().size());
+	}
+
+	public void testDecode003() throws IOException {
+		CashKyoroFile base = new CashKyoroFile(512);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *base*. Currently, there are execution paths that do not contain closure statements, for example, when CashKyoroFile.addChunk throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).